pkg/operator/controllers/pullsecret/pullsecret_controller.go

```diff
@@ -162,3 +135,90 @@ func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Named(controllers.PullSecretControllerName).
 		Complete(r)
 }
+
+// fixAndUpdateGlobalPullSecret checks the state of the pull secrets, in case of missing or broken ARO pull secret
+// it replaces it with working one from controller Secret
+func (r *PullSecretReconciler) fixAndUpdateGlobalPullSecret(ctx context.Context, operatorSecret, userSecret *corev1.Secret) (secret *corev1.Secret, err error) {
+	if operatorSecret == nil {
+		return nil, errors.New("nil operator secret, cannot verify userData integrity")
+	}
+
+	create := userSecret == nil
+	remove := false
+
+	// userSecret can happen to have broken type, which means it have to be recreated
+	// with proper type
+	if userSecret != nil && (userSecret.Type != corev1.SecretTypeDockerConfigJson || userSecret.Data == nil) {
+		create = true
+		remove = true
+	}
+
+	if create {
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      pullSecretName.Name,
+				Namespace: pullSecretName.Namespace,
+			},
+			Type: corev1.SecretTypeDockerConfigJson,
+		}
+		secret.Data = make(map[string][]byte)
+	} else {
+		secret = userSecret.DeepCopy()
+		if !json.Valid(secret.Data[corev1.DockerConfigJsonKey]) {
+			delete(secret.Data, corev1.DockerConfigJsonKey)
+		}
+	}
+
+	fixedData, update, err := pullsecret.Merge(string(secret.Data[corev1.DockerConfigJsonKey]), string(operatorSecret.Data[corev1.DockerConfigJsonKey]))
+	if err != nil {
+		return nil, err
+	}
+
+	if !create && !update {
+		return userSecret, nil
+	}
+
+	if remove {
+		// unfortunately the type field is immutable.
```

Comment:
should this comment be somewhere near line 150?
Or is it the only scenario when the secret will be removed ever?

Reply:
you are right, this make sense to mention on the line where it is set. The recreation is required when the `.type` field is broken. This field is immutable. I have moved the comment.